File: statescript/executor.go
// Copyright 2019 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

package statescript

import (
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"syscall"
	"time"

	"github.com/mendersoftware/log"
	"github.com/mendersoftware/mender/client"
	"github.com/pkg/errors"
)

const (
	// exitRetryLater - exit code returned if a script requests a retry
	exitRetryLater = 21

	defaultStateScriptRetryInterval time.Duration = 60 * time.Second

	defaultStateScriptRetryTimeout time.Duration = 30 * time.Minute

	defaultStateScriptTimeout time.Duration = 1 * time.Hour
)

type Executor interface {
	ExecuteAll(state, action string, ignoreError bool, report *client.StatusReportWrapper) error
	CheckRootfsScriptsVersion() error
}

type Launcher struct {
	ArtScriptsPath          string
	RootfsScriptsPath       string
	SupportedScriptVersions []int
	Timeout                 int
	RetryInterval           int
	RetryTimeout            int
}

func (l *Launcher) getRetryInterval() time.Duration {

	if l.RetryInterval != 0 {
		return time.Duration(l.RetryInterval) * time.Second
	}
	log.Warningf("No timeout interval set for the retry-scripts. Falling back to default: %s", defaultStateScriptRetryInterval.String())
	return defaultStateScriptRetryInterval
}

func (l *Launcher) getRetryTimeout() time.Duration {

	if l.RetryTimeout != 0 {
		return time.Duration(l.RetryTimeout) * time.Second
	}
	log.Warningf("No total time set for the retry-scripts' timeslot. Falling back to default: %s", defaultStateScriptRetryTimeout.String())
	return defaultStateScriptRetryTimeout
}

func (l Launcher) getTimeout() time.Duration {

	if l.Timeout != 0 {
		return time.Duration(l.Timeout) * time.Second
	}
	log.Debugf("statescript: timeout for executing scripts is not defined; using default of %s seconds", defaultStateScriptTimeout)
	return defaultStateScriptTimeout
}

//TODO: we can optimize for reading directories once and then creating
// a map with all the scripts that needs to be executed.
func (l Launcher) CheckRootfsScriptsVersion() error {
	// first check if we are having some scripts
	scripts, err := ioutil.ReadDir(l.RootfsScriptsPath)
	if err != nil && os.IsNotExist(err) {
		// no scripts; no error
		return nil
	} else if err != nil {
		return errors.Wrap(err, "statescript: can not read rootfs scripts directory")
	}

	if len(scripts) == 0 {
		return nil
	}

	versionFilePath := filepath.Join(l.RootfsScriptsPath, "version")
	f, err := os.Open(versionFilePath)
	if err != nil && os.IsNotExist(err) {
		errmsg := "statescript: The statescript version file is missing. This file needs to be " +
			"present and contain a single number representing which version of the statescript " +
			"support the client is using in order to successfully run statescripts on the client"
		return errors.New(errmsg)
	} else if err != nil {
		return errors.Wrap(err, "statescript")
	}
	ver, err := readVersion(f)
	if _, ok := err.(readVersionParseError); ok {
		errmsg := "statescript: Failed to parse the version file in the statescript directory (%s). " +
			"The file needs to contain a single integer signifying which version of the statescript " +
			"support which this client is using"
		return fmt.Errorf(errmsg, err)
	}
	if err != nil {
		return errors.Wrap(err, "statescript")
	}

	for _, v := range l.SupportedScriptVersions {
		if v == ver {
			return nil
		}
	}
	return errors.Errorf("statescript: unsupported scripts version: %v", ver)
}

func matchVersion(actual int, supported []int, hasScripts bool) error {
	// if there are no scripts to execute we shold not care about the version
	if hasScripts == false {
		return nil
	}

	for _, v := range supported {
		if v == actual {
			return nil
		}
	}

	errmsg := "statescript: The version read from the version file in the statescript directory " +
		"does not match the versions supported by the client, please make sure the file is present " +
		"and formatted correctly (supported: %v; actual: %v)."
	return errors.Errorf(errmsg, supported, actual)
}

func (l Launcher) get(state, action string) ([]os.FileInfo, string, error) {

	sDir := l.ArtScriptsPath
	if state == "Idle" || state == "Sync" || state == "Download" {
		sDir = l.RootfsScriptsPath
	}

	// ReadDir reads the directory named by dirname and returns
	// a list of directory entries sorted by filename.
	// The list returned should be sorted which guarantees correct
	// order of scripts execution.
	files, err := ioutil.ReadDir(sDir)
	if err != nil && os.IsNotExist(err) {
		// no state scripts directory; just move on
		return nil, "", nil
	} else if err != nil {
		return nil, "", errors.Wrap(err, "statescript: can not read scripts directory")
	}

	scripts := make([]os.FileInfo, 0)
	var version int

	for _, file := range files {
		if file.Name() == "version" {
			f, err := os.Open(filepath.Join(sDir, file.Name()))
			if err != nil {
				return nil, "", errors.Wrapf(err, "statescript")
			}
			version, err = readVersion(f)
			if err != nil {
				return nil, "", errors.Wrapf(err, "statescript")
			}
		}

		if strings.Contains(file.Name(), state+"_") &&
			strings.Contains(file.Name(), action) {

			// all scripts must be formated like `ArtifactInstall_Enter_05(_wifi-driver)`(optional)
			re := regexp.MustCompile(`([A-Za-z]+)_(Enter|Leave|Error)_[0-9][0-9](_\S+)?`)
			if len(file.Name()) == len(re.FindString(file.Name())) {
				scripts = append(scripts, file)
			} else {
				log.Warningf("script format mismatch: '%s' will not be run ", file.Name())
			}
		}
	}

	if err := matchVersion(version, l.SupportedScriptVersions,
		len(scripts) != 0); err != nil {
		return nil, "", err
	}

	return scripts, sDir, nil
}

func execute(name string, timeout time.Duration) error {

	cmd := exec.Command(name)

	var stderr io.ReadCloser
	var err error

	if !strings.HasPrefix(name, "Idle") && !strings.HasPrefix(name, "Sync") {
		stderr, err = cmd.StderrPipe()
		if err != nil {
			log.Errorf("statescript: %v", err)
			return errors.Wrap(err, "statescript: unable to open stderr pipe")
		}
	}

	// As child process gets the same PGID as the parent by default, in order
	// to avoid killing Mender when killing process group we are setting
	// new PGID for the executed script and its children.
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}

	if err := cmd.Start(); err != nil {
		return err
	}

	timer := time.AfterFunc(timeout, func() {
		// In addition to kill a single process we are sending SIGKILL to
		// process group making sure we are killing the hanging script and
		// all its children.
		syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
	})
	defer timer.Stop()

	var bts []byte
	if stderr != nil {
		bts, err = ioutil.ReadAll(stderr)
		if err != nil {
			log.Error(err)
		}
	}

	if len(bts) > 0 {
		if len(bts) > 10*1024 {
			log.Infof("Collected output (stderr) while running script %s (Truncated to 10KB)\n%s\n---------- end of script output", name, bts[:10*1024])
		} else {
			log.Infof("Collected output (stderr) while running script %s\n%s\n---------- end of script output", name, string(bts))
		}
	}

	if err := cmd.Wait(); err != nil {
		return err
	}
	return nil
}

func retCode(err error) int {
	defaultFailedCode := -1

	if err != nil {
		// try to get the exit code
		if exitError, ok := err.(*exec.ExitError); ok {
			ws := exitError.Sys().(syscall.WaitStatus)
			return ws.ExitStatus()
		} else {
			return defaultFailedCode
		}
	}
	return 0
}

// Catches a script that requests a retry in a loop. Is limited by the total window given to a script demanding a
// retry.
func executeScript(s os.FileInfo, dir string, l Launcher, timeout time.Duration, ignoreError bool) error {

	iet := time.Now()
	for {
		err := execute(filepath.Join(dir, s.Name()), timeout)
		switch ret := retCode(err); ret {
		case 0:
			// success
			return nil
		case exitRetryLater:
			if time.Since(iet) <= l.getRetryTimeout() {
				log.Infof("statescript: %s requested a retry", s.Name())
				time.Sleep(l.getRetryInterval())
				continue
			}
			if ignoreError {
				log.Errorf("statescript: ignoring error executing '%s': %d: %s", s.Name(), ret, err.Error())
				return nil
			}
			return errors.Errorf("statescript: retry time-limit exceeded %s", err.Error())
		default:
			// In case of error scripts all should be executed.
			if ignoreError {
				log.Errorf("statescript: ignoring error executing '%s': %d: %s", s.Name(), ret, err.Error())
				return nil
			}
			return errors.Errorf("statescript: error executing '%s': %d : %s",
				s.Name(), ret, err.Error())
		}
	}
}

func reportScriptStatus(rep *client.StatusReportWrapper, statusReport string) error {
	c := client.NewStatus()

	return c.Report(rep.API, rep.URL, client.StatusReport{
		DeploymentID: rep.Report.DeploymentID,
		Status:       rep.Report.Status,
		SubState:     statusReport,
	})
}

func (l Launcher) ExecuteAll(state, action string, ignoreError bool,
	report *client.StatusReportWrapper) error {
	scr, dir, err := l.get(state, action)
	if err != nil {
		if ignoreError {
			log.Errorf("statescript: got an error when trying to execute [%s:%s] script, "+
				"but ignoreError is set to true, so continuing. Full error message: %v",
				state, action, err)
			return nil
		}
		return err
	}

	execBits := os.FileMode(syscall.S_IXUSR | syscall.S_IXGRP | syscall.S_IXOTH)
	timeout := l.getTimeout()

	for _, s := range scr {
		// check if script is executable
		if s.Mode()&execBits == 0 {
			if ignoreError {
				log.Errorf("statescript: ignoring script '%s' being not executable",
					filepath.Join(dir, s.Name()))
				continue
			} else {
				return errors.Errorf("statescript: script '%s' is not executable",
					filepath.Join(dir, s.Name()))
			}
		}

		subStatus := fmt.Sprintf("start executing script: %s", s.Name())
		log.Debugf(subStatus)
		if report != nil {
			if err = reportScriptStatus(report, subStatus); err != nil {
				log.Errorf("statescript: can not send start status to server: %s", err.Error())
			}

			defer func() {
				if err = reportScriptStatus(report,
					fmt.Sprintf("finished executing script: %s", s.Name())); err != nil {
					log.Errorf("statescript: can not send finished status to server: %s", err.Error())
				}
			}()
		}

		if err = executeScript(s, dir, l, timeout, ignoreError); err != nil {
			return err
		}
	}
	return nil
}

File: artifact/v3/scripts/executor.cpp
// Copyright 2023 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

#include <artifact/v3/scripts/executor.hpp>

#include <algorithm>
#include <chrono>
#include <regex>
#include <string>

#include <common/common.hpp>
#include <common/expected.hpp>
#include <common/path.hpp>


namespace mender {
namespace artifact {
namespace scripts {
namespace executor {

namespace expected = mender::common::expected;


using expected::ExpectedBool;

namespace common = mender::common;
namespace error = mender::common::error;
namespace path = mender::common::path;


const string expected_state_script_version {"3"};
const int state_script_retry_exit_code {21};

unordered_map<const State, string> state_map {
	{State::Idle, "Idle"},
	{State::Sync, "Sync"},
	{State::Download, "Download"},
	{State::ArtifactInstall, "ArtifactInstall"},
	{State::ArtifactReboot, "ArtifactReboot"},
	{State::ArtifactCommit, "ArtifactCommit"},
	{State::ArtifactRollback, "ArtifactRollback"},
	{State::ArtifactRollbackReboot, "ArtifactRollbackReboot"},
	{State::ArtifactFailure, "ArtifactFailure"},
};

unordered_map<const Action, string> action_map {
	{Action::Enter, "Enter"},
	{Action::Leave, "Leave"},
	{Action::Error, "Error"},
};

error::Error CorrectVersionFile(const string &path) {
	// Missing file is OK
	// This is because previous versions of the client wrote no
	// version file, so no-file=v3
	if (!path::FileExists(path)) {
		return error::NoError;
	}

	ifstream vf {path};

	if (!vf) {
		auto errnum {errno};
		return error::Error(
			generic_category().default_error_condition(errnum), "Failed to open the version file");
	}

	string version;
	vf >> version;
	if (!vf) {
		auto errnum {errno};
		return error::Error(
			generic_category().default_error_condition(errnum),
			"Error reading the version number from the version file");
	}

	if (version != expected_state_script_version) {
		return executor::MakeError(
			executor::VersionFileError, "Unexpected Artifact script version found: " + version);
	}
	return error::NoError;
}

bool isValidStateScript(const string &file, State state, Action action) {
	string expression {
		"(" + state_map.at(state) + ")" + "_(" + action_map.at(action) + ")_[0-9][0-9](_\\S+)?"};
	log::Trace(
		"verifying the State script format of the file: " + file
		+ " using the regular expression: " + expression);
	const regex artifact_script_regexp {expression, std::regex_constants::ECMAScript};
	return regex_match(path::BaseName(file), artifact_script_regexp);
}

function<bool(const string &)> Matcher(State state, Action action) {
	return [state, action](const string &file) {
		const bool is_valid {isValidStateScript(file, state, action)};
		if (!is_valid) {
			log::Trace(file + " is not a valid State Script for the state: " + Name(state, action));
			return false;
		}
		auto exp_executable = path::IsExecutable(file, true);
		if (!exp_executable) {
			log::Debug("Issue figuring the executable bits of: " + exp_executable.error().String());
			return false;
		}
		return is_valid and exp_executable.value();
	};
}

bool IsArtifactScript(State state) {
	switch (state) {
	case State::Idle:
	case State::Sync:
	case State::Download:
		return false;
	case State::ArtifactInstall:
	case State::ArtifactReboot:
	case State::ArtifactCommit:
	case State::ArtifactRollback:
	case State::ArtifactRollbackReboot:
	case State::ArtifactFailure:
		return true;
	}
	assert(false);
	return false;
}

string ScriptRunner::ScriptPath(State state) {
	if (IsArtifactScript(state)) {
		return this->artifact_script_path_;
	}
	return this->rootfs_script_path_;
}

string Name(const State state, const Action action) {
	return state_map.at(state) + action_map.at(action);
}

ScriptRunner::ScriptRunner(
	events::EventLoop &loop,
	chrono::seconds state_script_timeout,
	const string &artifact_script_path,
	const string &rootfs_script_path,
	mender::common::processes::OutputCallback stdout_callback,
	mender::common::processes::OutputCallback stderr_callback) :
	loop_ {loop},
	state_script_timeout_ {state_script_timeout},
	artifact_script_path_ {artifact_script_path},
	rootfs_script_path_ {rootfs_script_path},
	stdout_callback_ {stdout_callback},
	stderr_callback_ {stderr_callback},
	error_script_error_ {error::NoError} {};

void ScriptRunner::LogErrAndExecuteNext(
	Error err,
	vector<string>::iterator current_script,
	vector<string>::iterator end,
	bool ignore_error,
	HandlerFunction handler) {
	// Collect the error and carry on
	if (err.code
		== common::processes::MakeError(common::processes::NonZeroExitStatusError, "").code) {
		this->error_script_error_ = this->error_script_error_.FollowedBy(executor::MakeError(
			executor::NonZeroExitStatusError,
			"Got non zero exit code from script: " + *current_script));
	} else {
		this->error_script_error_ = this->error_script_error_.FollowedBy(err);
	}

	// Schedule the next script execution
	auto local_err = Execute(std::next(current_script), end, ignore_error, handler);
	if (local_err != error::NoError) {
		return handler(local_err);
	}
}

void ScriptRunner::HandleScriptError(Error err, HandlerFunction handler) {
	if (err.code
		== common::processes::MakeError(common::processes::NonZeroExitStatusError, "").code) {
		if (this->script_.get()->GetExitStatus() == state_script_retry_exit_code) {
			return handler(executor::MakeError(
				executor::RetryExitCodeError,
				"Received exit code: " + to_string(state_script_retry_exit_code)));
		}
		return handler(executor::MakeError(
			executor::NonZeroExitStatusError,
			"Received error code: " + to_string(this->script_.get()->GetExitStatus())));
	}
	return handler(err);
}

Error ScriptRunner::Execute(
	vector<string>::iterator current_script,
	vector<string>::iterator end,
	bool ignore_error,
	HandlerFunction handler) {
	// No more scripts to execute
	if (current_script == end) {
		handler(this->error_script_error_); // Success
		return error::NoError;
	}

	log::Info("Running State Script: " + *current_script);

	this->script_.reset(new mender::common::processes::Process({*current_script}));
	auto err {this->script_->Start(stdout_callback_, stderr_callback_)};
	if (err != error::NoError) {
		return err;
	}

	return this->script_.get()->AsyncWait(
		this->loop_,
		[this, current_script, end, ignore_error, handler](Error err) {
			if (err != error::NoError) {
				if (ignore_error) {
					return LogErrAndExecuteNext(err, current_script, end, ignore_error, handler);
				}
				return HandleScriptError(err, handler);
			}
			// Schedule the next script execution
			auto local_err = Execute(std::next(current_script), end, ignore_error, handler);
			if (local_err != error::NoError) {
				return handler(local_err);
			}
		},
		this->state_script_timeout_);
}

Error ScriptRunner::AsyncRunScripts(
	State state, Action action, HandlerFunction handler, RunError on_error) {
	if (IsArtifactScript(state)) {
		// Verify the version in the version file (OK if no version file present)
		auto version_file_error {
			CorrectVersionFile(path::Join(this->artifact_script_path_, "version"))};
		if (version_file_error != error::NoError) {
			return version_file_error;
		}
	}

	// Collect
	const auto script_path {ScriptPath(state)};
	auto exp_scripts {path::ListFiles(script_path, Matcher(state, action))};
	if (!exp_scripts) {
		// Missing directory is OK
		if (exp_scripts.error().IsErrno(ENOENT)) {
			log::Warning("Found no state script directory (" + script_path + "). Continuing on");
			handler(error::NoError);
			return error::NoError;
		}
		return executor::MakeError(
			executor::Code::CollectionError,
			"Failed to get the scripts, error: " + exp_scripts.error().String());
	}

	// Sort
	{
		auto &unsorted_scripts {exp_scripts.value()};

		vector<string> sorted_scripts(unsorted_scripts.begin(), unsorted_scripts.end());

		sort(sorted_scripts.begin(), sorted_scripts.end());
		this->collected_scripts_ = std::move(sorted_scripts);
	}

	bool ignore_error = on_error == RunError::Ignore || action == Action::Error;

	// Execute
	auto scripts_iterator {this->collected_scripts_.begin()};
	auto scripts_iterator_end {this->collected_scripts_.end()};
	return Execute(scripts_iterator, scripts_iterator_end, ignore_error, handler);
}


Error ScriptRunner::RunScripts(State state, Action action, RunError on_error) {
	auto run_err {error::NoError};
	auto err = AsyncRunScripts(
		state,
		action,
		[this, &run_err](Error error) {
			run_err = error;
			this->loop_.Stop();
		},
		on_error);
	if (err != error::NoError) {
		return err;
	}
	this->loop_.Run();
	return run_err;
}


} // namespace executor
} // namespace scripts
} // namespace artifact
} // namespace mender
